---
title: "Reproducibiliy sample type"
output: html_document
---


```{r setup, message=FALSE, warning=FALSE}
library(tidyverse)
library(tableone)
library(ggplot2)
library(ggpubr)
library(cowplot)
library(here)
library(table1)


```


```{r import_data}
sample_info <-
  read_csv(here("raw_data", "Samples_RBC_DBS.csv"), col_types = cols())  %>% 
  select(sample_id = SampleID, patient_id = patient_id, sample_type = sample_type, sample_day = sample_day, visit_day = visit_day) %>% 
  distinct()

haplotype_calls <-
  read_csv(here("raw_data","Haplotype_calls_RBC_DBS.csv"), col_types = cols())  %>% 
  select(marker_id = MarkerID, sample_id = SampleID, sample_name = SampleName, haplotype = Haplotype, n_reads = Reads, freq = Frequency) %>% 
  mutate(haplotype = case_when(haplotype == 'Noise'     ~ 'noise',
                               haplotype == 'Indels'    ~ 'indel',
                               haplotype == 'Singelton' ~ 'singleton',
                               haplotype == 'Chimera'   ~ 'chimera',
                               TRUE                     ~  haplotype)) %>% 
  replace_na(list(freq = 0))
```

```{r Retrieving data from AmpSeqR, message=FALSE, warning=FALSE}
SI_haplotypes <- 
  haplotype_calls %>% 
  left_join(sample_info) %>% 
  group_by(marker_id, sample_id,patient_id,sample_day,visit_day,sample_type,haplotype) %>%   summarise(n_reads = mean(n_reads)) %>% 
   group_by(marker_id, sample_id) %>% 
   filter(! haplotype %in% c('indel', 'singleton', 'chimera', 'noise', NA),n_reads >= 10) %>%
  group_by(marker_id, sample_id, patient_id,sample_day,visit_day, sample_type, haplotype, n_reads) %>%   count() 

#Editing naming of haplotypes of chr 07-after aligning and checking the identity between haplotype names
SI_hap_edited <- SI_haplotypes %>%
  ungroup(haplotype, marker_id) %>%
      mutate(haplotype = case_when(haplotype == "Chr07_G-1" ~ "Chr07-1", 
                                  haplotype == "Chr07_G-2"  ~ "Chr07-2", 
                                  haplotype ==  "Chr07_G-3" ~ "Chr07-3",
                                  haplotype == "Chr07_G-4" ~ "Chr07-4",
                                  haplotype == "Chr07_G-5" ~ "Chr07-5",
                                  haplotype == "Chr07_G-6" ~ "Chr07-6", 
                                  haplotype == "Chr07_G-8" ~ "Chr07-8", 
                                  haplotype == "Chr07_G-9" ~ "Chr07-9", 
                                  haplotype ==  "Chr07_G-11" ~ "Chr07-11", 
                                  haplotype == "Chr07_G-12" ~ "Chr07-12",
                                  TRUE ~ haplotype ),
            marker_id = case_when(marker_id == "Chr07_G"  ~ "Chr07",
                                  TRUE ~ marker_id)) %>%
    filter(! marker_id %in% c('Chr07_A'), ! sample_type  %in% c("control") )

#######################################################
#   Comparison of agreement between RBC and DBS data  #
#######################################################

##############################
#  Haplotypes >= 1% in-host  #
##############################

# Splitting data of duplicates at day 0: DBS (n = 20) and RBC (n = 22)

SI_RBC_day0 <- SI_hap_edited[c(SI_hap_edited$sample_type == "RBC" & SI_hap_edited$visit_day == 0),] %>% group_by(patient_id,marker_id,haplotype, sample_type) %>%  summarise(total_reads = sum(n_reads)) %>%  mutate(freq = total_reads/sum(total_reads)) %>% group_by(patient_id) %>% filter(freq >= 0.01)

SI_DBS_day0 <- SI_hap_edited[c(SI_hap_edited$sample_type == "DBS" & SI_hap_edited$visit_day == 0),] %>% group_by(patient_id,marker_id,haplotype,sample_type) %>%  summarise(total_reads = sum(n_reads)) %>%  mutate(freq = total_reads/sum(total_reads)) %>% group_by(patient_id) %>% filter(freq >= 0.01)

#Set of patient_id's with both RBC and DBS samples at day 0 

RBC_day0_list <- SI_RBC_day0 %>% select(patient_id, sample_type) %>% unique()
DBS_day0_list <- SI_DBS_day0 %>% select(patient_id, sample_type) %>% unique()

RBC_and_DBS_samples_day0_list <- RBC_day0_list %>% 
  filter(patient_id %in% DBS_day0_list$patient_id)

```

```{r Correlation Parasite density and number of reads}


RBC_and_DBS_reads_coverage<- SI_RBC_day0 %>%
  bind_rows(SI_DBS_day0) %>% select(!c(haplotype,freq)) %>%
  filter(patient_id %in% RBC_and_DBS_samples_day0_list$patient_id) %>% 
  group_by(patient_id, sample_type, marker_id) %>%
mutate(reads = (sum(total_reads))) %>% select(patient_id, sample_type,marker_id, reads) %>%
  distinct()

n_distinct(RBC_and_DBS_reads_coverage)


table1(~patient_id + marker_id|sample_type, data =  RBC_and_DBS_reads_coverage, 
       render.missing = NULL,
       render.categorical = "FREQ (PCTnoNA%)")
       #render = rndr)

table1(~reads |marker_id + sample_type, data =  RBC_and_DBS_reads_coverage, 
       render.missing = NULL,
       render.categorical = "FREQ (PCTnoNA%)")

ptest <- RBC_and_DBS_reads_coverage %>% group_by(marker_id) %>% summarize(p.value = kruskal.test(reads ~sample_type)$p.value)

#Boxplot
RBC_DBS_reads_boxplot<-
  RBC_and_DBS_reads_coverage %>%
  ggplot(aes(x = sample_type, y = log10(reads), fill=marker_id)) + 
  geom_boxplot(outlier.shape = NA, alpha = 0.5) + 
 geom_jitter(alpha = 0.5, width = 0.1, size = 3)+ 
  geom_text(data = ptest, aes(x =  c(1.5,1.5,1.5,1.5,1.5,1.5), 
  y = c(6,6,6,6,6,6), label = paste0("p=",round(p.value,3)))) +
  scale_y_log10()+ 
  ggtitle("Read coverage from DBS and RBC (17 paired samples)") + 
  ylab("log10(Reads)") + xlab("Sample type") +
  theme_bw(base_size = 15) + scale_fill_brewer(palette="Paired") +
  theme(axis.text.x = element_text(face = "plain", color="black", size = 15, angle = 0), legend.position = "none")+ 
  facet_wrap(~ marker_id, ncol = 6)  

print(RBC_DBS_reads_boxplot)


```






```{r Plotting matching RBC and DBS haplotypes at day 0}

Compare_DBS_RBC_MOI <- 
  RBC_and_DBS_samples_day0 %>%
  ungroup() %>%
  group_by(patient_id, marker_id) %>%
  mutate(MOI_RBC = n_distinct(haplotype.x),
         MOI_DBS = n_distinct(haplotype.y)) %>%
  ungroup() %>%
  select(marker_id,MOI_RBC,MOI_DBS) %>%
  group_by(marker_id) %>%
  pivot_longer(!marker_id,
    names_to = "sample_MOI",
    values_to = "MOI"
  )%>%
  mutate(MOI = as.factor(MOI)) %>%
  group_by(marker_id, sample_MOI, MOI) %>%
  mutate(sample_MOI = factor(sample_MOI, 
                        levels = c("MOI_DBS","MOI_RBC"), 
                      labels = c("DBS","RBC"))) %>%
  count()



DBS_RBC_MOI_plot <- Compare_DBS_RBC_MOI %>%
  ggplot( ) +
  geom_col(aes(x = sample_MOI, y = n,fill = MOI), col = 'black', position = "fill") +
  ylab(" MOI proportion") + xlab(" Sample type ") +  
  ggtitle("Multiplicity of infection from DBS and RBC (17 paired samples) ") +
  theme_classic(base_size = 15) + 
  theme(axis.text.x = element_text(face = "plain", color="black", size = 15, angle = 0), legend.position = "bottom")+
  facet_wrap(~marker_id, ncol = 6)

print(DBS_RBC_MOI_plot)  

Supp_figure2 <- ggdraw() +
  draw_plot(RBC_DBS_reads_boxplot, x = 0.005,y = 0.5, width = .89, height = 0.47) +
  draw_plot(DBS_RBC_MOI_plot, x = 0.005,y = 0, width = .89, height = 0.47) +
  draw_plot_label(label = c("A", "B"), size = 12,
                 x = c(0,0), y = c(1, 0.5))

print(Supp_figure2)


ggsave("figs/Supp_figure1.tiff", units="cm", width=25, height=25, dpi=300, compression = 'lzw')



```

Descriptive table
```{r}

MOI_stats<- 
RBC_and_DBS_samples_day0 %>%
  ungroup() %>%
  group_by(patient_id, marker_id) %>%
  mutate(MOI_RBC = n_distinct(haplotype.x),
         MOI_DBS = n_distinct(haplotype.y)) %>%
  ungroup() %>%
  select(marker_id,MOI_RBC,MOI_DBS) %>%
  group_by(marker_id) %>%
  pivot_longer(!marker_id,
    names_to = "sample_MOI",
    values_to = "MOI" )


table_DBS_RBC<- CreateTableOne(vars = c("MOI"), strata = "sample_MOI", data = MOI_stats[MOI_stats$marker_id=="Chr03",], 
               factorVars =c("marker_id"))

CreateTableOne(vars = c("MOI"), strata = "sample_MOI", data = MOI_stats[MOI_stats$marker_id=="Chr05",], 
               factorVars =c("marker_id"))

CreateTableOne(vars = c("MOI"), strata = "sample_MOI", data = MOI_stats[MOI_stats$marker_id=="Chr06",], 
               factorVars =c("marker_id"))

CreateTableOne(vars = c("MOI"), strata = "sample_MOI", data = MOI_stats[MOI_stats$marker_id=="Chr07",], 
               factorVars =c("marker_id"))

CreateTableOne(vars = c("MOI"), strata = "sample_MOI", data = MOI_stats[MOI_stats$marker_id=="Chr08",], 
               factorVars =c("marker_id"))

CreateTableOne(vars = c("MOI"), strata = "sample_MOI", data = MOI_stats[MOI_stats$marker_id=="Chr13",], 
               factorVars =c("marker_id"))

```





