---
title: "SNP number_Peru_all_samples"
output:
  html_document:
    df_print: paged
---

**R packages**

```{r,warning=FALSE,message=FALSE}
library(tidyverse)
library(DECIPHER)
library(Biostrings)
library(ShortRead)
library(here)
```


```{r echo=FALSE, message=F, warning=F}
load("./intermediate_data/intermediate_data.RData")

```

**SNPs**

```{r}
#Data
data_all <- hap_PE_epi %>%
 dplyr::select(sample_id, marker_id,sequence)

marker_info <- read.csv(here("raw_data","marker_info.csv"))
# SNPs
# Get the variant position
all_markers <- marker_info
all_markers <- all_markers %>%
  filter(marker_id %in% as.character(data_all$marker_id))
seq_tbl_all <- data_all %>%
    mutate(seq_id = map_chr(sequence, digest::digest))
#########################################function#########################################
#call variants function
call_vars <- function(seq_tbl, marker_info) {
  proc_vars <- process_vars(seq_tbl, marker_info)

  vars_tbl <-
    proc_vars %>%
    mutate(var_data = map(var_data, "vars")) %>%
    unnest(var_data) %>%
    mutate(var_id = str_c(marker_id, var_id, sep = "_")) %>%
    select(marker_id, var_id, everything())

  return(vars_tbl)
}

#process variants function
process_vars <- function(seq_tbl, marker_info) {

  # check args
  if (!all(seq_tbl$marker_id %in% marker_info$marker_id)) {
    abort("not all marker_ids in asv_tbl present in marker_info")
  }

  seq_tbl %>%
    select(marker_id, sequence, seq_id) %>%
    na.omit() %>%
    distinct() %>%
    chop(-marker_id) %>%
    mutate(dnass = map2(seq_id, sequence, function(aid, seq) {
      DNAStringSet(seq) %>% setNames(aid)
    })) %>%
    inner_join(select(marker_info, marker_id, marker_seq = seq), "marker_id") %>%
    mutate(var_data = map2(marker_seq, dnass, function(marker_seq, seqs) {
      call_variants(DNAStringSet(marker_seq), seqs)
    })) %>%
    select(marker_id, var_data)
}
#call variants function
call_variants <- function(ref_seq, var_seqs) {

  n <- length(var_seqs)

  aln_tbl <-
    DECIPHER::AlignSeqs(c(setNames(ref_seq, "ref"), var_seqs), verbose = FALSE) %>%
    as.matrix() %>%
    t() %>%
    as_tibble() %>%
    mutate(
      pos = cumsum(ref != "-"),
      aln_pos = seq_along(pos)
    ) %>%
    pivot_longer(c(-ref, -pos, -aln_pos),
      names_to = "seq_id",
      values_to = "asv_base"
    ) %>%
    (function(x) {
      tibble(ref = "N", pos = 0L, aln_pos = 0L, asv_base = "N", seq_id = unique(x$seq_id)) %>%
        bind_rows(x)
    }) %>%
    arrange(seq_id, aln_pos)


  vars <- if (length(unique(aln_tbl$seq_id)) == 1 & all(aln_tbl$ref == aln_tbl$asv_base)) {
    return(NULL)
    aln_tbl <- NULL
  } else {
    aln_tbl %>%
      split.data.frame(.$seq_id) %>%
      map_df(function(data) {
        data %>%
          filter(!(ref == "-" & asv_base == "-")) %>%
          mutate(
            ins_grp_start = replace_na(lead(ref) == "-" & ref != "-", FALSE),
            ins_grp_end = replace_na(ref != "-" & lag(ref) == "-", FALSE),
            ins_id = if_else(cumsum(ins_grp_start) > cumsum(ins_grp_end),
              cumsum(ins_grp_start),
              NA_integer_
            ),
            del_grp_start = replace_na(lead(asv_base) == "-" & asv_base != "-", FALSE),
            del_grp_end = replace_na(asv_base != "-" & lag(asv_base) == "-", FALSE),
            del_id = if_else(cumsum(del_grp_start) > cumsum(del_grp_end),
              cumsum(del_grp_start),
              NA_integer_
            ),
            is_snp = is.na(ins_id) & is.na(del_id) & ref != asv_base,
            snp_id = if_else(is_snp,
              cumsum(is_snp),
              NA_integer_
            )
          ) %>%
          select(-ins_grp_start, -ins_grp_end, -del_grp_start, -del_grp_end, -is_snp) %>%
          filter(!is.na(snp_id) | !is.na(ins_id) | !is.na(del_id)) %>%
          mutate(type = case_when(
            !is.na(snp_id) ~ "SNP",
            !is.na(ins_id) ~ "INS",
            !is.na(del_id) ~ "DEL"
          )) %>%
          group_by(ins_id, del_id, snp_id) %>%
          summarise(
            aln_pos = aln_pos[1],
            pos = pos[1],
            ref = str_c(ref, collapse = "") %>% str_remove_all("-"),
            asv_base = str_c(asv_base, collapse = "") %>% str_remove_all("-"),
            type = type[1],
            .groups = "drop"
          ) %>%
          mutate(seq_id = data$seq_id[1]) %>%
          select(seq_id, aln_pos, pos, ref, alt = asv_base, type)
      }) %>%
      chop(c(seq_id)) %>%
      arrange(aln_pos, pos, ref, alt) %>%
      chop(c(-aln_pos, -pos, -ref, -type)) %>%
      mutate(genotype = map(alt, seq_along)) %>%
      unnest(c(seq_id, genotype)) %>%
      unnest(seq_id) %>%
      group_by(aln_pos, pos, ref, alt, type) %>%
      complete(seq_id = names(var_seqs), fill = list(genotype = 0L)) %>%
      ungroup() %>%
      mutate(num_allele = lengths(alt) + 1L) %>%
      nest(gt_data = c(seq_id, genotype)) %>%
      mutate(var_id = str_c(pos, ref, type, sep = "_")) %>%
      select(var_id, everything()) %>%
      arrange(pos, var_id)
  }


  # find del sites to replace allele with '*'
  if (any(vars$type == "DEL")) {
    fixed <-
      vars %>%
      filter(type == "DEL") %>%
      mutate(end = pos + nchar(ref) - 1L) %>%
      unnest(gt_data) %>%
      filter(genotype != 0) %>%
      select(pos, end, seq_id) %>%
      chop(seq_id) %>%
      pmap_df(function(pos, end, seq_id) {
        filter(vars, pos >= !!pos, pos <= !!end, type != "DEL") %>%
          (function(x) {
            if (nrow(x) > 0) {
              x %>%
                mutate(
                  alt = map(alt, ~ c(., "")) %>% vctrs::as_list_of(),
                  num_allele = num_allele + 1L,
                  gt_data = map2(gt_data, num_allele, function(data, na) {
                    mutate(data, genotype = if_else(seq_id %in% !!seq_id, na - 1L, genotype))
                  })
                )
            }
          })
      })

    if (nrow(fixed) > 0) {
      vars <-
        anti_join(vars, fixed, by = "var_id") %>%
        bind_rows(fixed) %>%
        arrange(pos, var_id)
    }
  }

  return(list(vars = vars, aln_tbl = aln_tbl))
}

#########################################call variants#########################################
#Use the call_vars function
vars <- call_vars(
    seq_tbl = seq_tbl_all,
    marker_info = all_markers
  )
vars <- vars %>%
    dplyr::select(marker_id, var_id, type)
#Get SNP number
snps <- vars %>%
    group_by(marker_id) %>%
    summarise(SNPs = n()) %>%
    ungroup()
snps

del <- vars %>%
    group_by(marker_id) %>%
  filter(type == "INS") %>%
    summarise(n = n())%>%
    #summarise(del = n()) %>%
    ungroup()



```

